File: C++/payment/paymentmanager.cpp
#include "paymentmanager.hpp"
#include "../config/config.hpp"

PaymentManager::PaymentManager() : paymentsolution(nullptr) {}

PaymentManager& PaymentManager::GetInstance() {
    static PaymentManager instance;
    return instance;
}

void PaymentManager::ResetPaymentSolution(const string& paymentProvider) {
    if (paymentProvider == "razorpay") {
        delete paymentsolution;
        paymentsolution = new RazorpayPaymentFactory();

    }
    else if (paymentProvider == "phonepe") {
        delete paymentsolution;
        paymentsolution = new PhonepePaymentFactory();

    }
    else {
        cout << "unknown payment provider . cannot reset ." << endl;
        return;
    }

    if (paymentsolution != nullptr) {
        paymentsolution = new RazorpayPaymentFactory();
        creditcardpayment = paymentsolution->CreateCreditCardPayment();
        netbanking = paymentsolution->CreateNetbankingPayment();
        upipayment = paymentsolution->CreateUPIPayment();
    }

}


void PaymentManager::Initialize() {
    //use config to initialize the payment solutions and methods
    ConfigReader& configreader =  ConfigReader::getInstance();
    auto configmap = configreader.GetConfigMap();
    auto sectionIter = configmap.find("payment");

    if (sectionIter != configmap.end()) {
        auto keyIterrazorpay = sectionIter->second.find("razorpay");
        auto keyIterphonepe = sectionIter->second.find("phonepe");

        if (keyIterrazorpay != sectionIter->second.end()) {
            bool value = keyIterrazorpay->second;
            if (value) {
                //similarly from the config check for each payment type and initialize
                paymentsolution = new RazorpayPaymentFactory();
                creditcardpayment =  paymentsolution->CreateCreditCardPayment();
                netbanking = paymentsolution->CreateNetbankingPayment();
                upipayment = paymentsolution->CreateUPIPayment();
            }
        }
        else if (keyIterphonepe != sectionIter->second.end()) {
            bool value = keyIterphonepe->second;
            if (value) {
                //similarly from the config check for each payment type and initialize
                paymentsolution = new PhonepePaymentFactory();
                creditcardpayment = paymentsolution->CreateCreditCardPayment();
                netbanking = paymentsolution->CreateNetbankingPayment();
                upipayment = paymentsolution->CreateUPIPayment();
            }

        }
        else {
            cout << "unknown payment solution . please check" << endl;
        }
    }
    else {
        cout << "config initialization failed" << endl;
    }
   
}

void PaymentManager::ProcessPaymentWithNetbanking(double amount) {
    netbanking->ProcessNetBankingPayment(amount);
}
void PaymentManager::ProcessPaymentWithCreditcard(double amount) {
    creditcardpayment->ProcessCreditCardPayment(amount);
}
void PaymentManager::ProcessPaymentWithUPI(double amount) {
    upipayment->ProcessUPIPayment(amount);
}

File: JAVA/Payment/PaymentManager.java
import java.util.*;
import config.ConfigReader;

public class PaymentManager {
    private static PaymentManager instance;
    private PaymentFactory paymentsolution;
    private NetBankingPayment netbanking;
    private CreditCardPayment creditcardpayment;
    private UPIPayment upipayment;

    private PaymentManager() {
        paymentsolution = null;
    }

    public static PaymentManager GetInstance() {
        if (instance == null) {
            instance = new PaymentManager();
        }
        return instance;
    }

    public void ResetPaymentSolution(String paymentProvider) {
        if (paymentProvider.equals("razorpay")) {
            paymentsolution = new RazorpayPaymentFactory();
        } else if (paymentProvider.equals("phonepe")) {
            paymentsolution = new PhonepePaymentFactory();
        } else {
            System.out.println("Unknown payment provider. Cannot reset.");
            return;
        }

        if (paymentsolution != null) {
            creditcardpayment = paymentsolution.CreateCreditCardPayment();
            netbanking = paymentsolution.CreateNetbankingPayment();
            upipayment = paymentsolution.CreateUPIPayment();
        }
    }

    public void Initialize() {
        // Use config to initialize the payment solutions and methods
        ConfigReader configreader = ConfigReader.getInstance();
        Map<String, Map<String, Boolean>> configmap = configreader.GetConfigMap();
        Map<String, Boolean> section = configmap.get("payment");

        if (section != null) {
            Boolean useRazorpay = section.get("razorpay");
            Boolean usePhonepe = section.get("phonepe");

            if (useRazorpay != null && useRazorpay) {
                paymentsolution = new RazorpayPaymentFactory();
                creditcardpayment = paymentsolution.CreateCreditCardPayment();
                netbanking = paymentsolution.CreateNetbankingPayment();
                upipayment = paymentsolution.CreateUPIPayment();
            } else if (usePhonepe != null && usePhonepe) {
                paymentsolution = new PhonepePaymentFactory();
                creditcardpayment = paymentsolution.CreateCreditCardPayment();
                netbanking = paymentsolution.CreateNetbankingPayment();
                upipayment = paymentsolution.CreateUPIPayment();
            } else {
                System.out.println("Unknown payment solution. Please check.");
            }
        } else {
            System.out.println("Config initialization failed.");
        }
    }

    public void ProcessPaymentWithNetbanking(double amount) {
        netbanking.ProcessNetBankingPayment(amount);
    }

    public void ProcessPaymentWithCreditcard(double amount) {
        creditcardpayment.ProcessCreditCardPayment(amount);
    }

    public void ProcessPaymentWithUPI(double amount) {
        upipayment.ProcessUPIPayment(amount);
    }
}
